inductor: fix _dynamic_reshape_indexer issue when tail index is sym

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97502


For TIMM **swin_base_patch4_window7_224**  dynamic shape case, there has an error for ```view``` op:

```
  File "/home/xiaobing/pytorch-offical/torch/_inductor/lowering.py", line 229, in wrapped
    out = decomp_fn(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/lowering.py", line 665, in view
    return TensorBox(View.create(x.data, sizes))
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1325, in create
    reindex = cls.dynamic_reshape_indexer(old_size, new_size)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1351, in dynamic_reshape_indexer
    reindex2 = cls._dynamic_reshape_indexer(flat, new_size)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1406, in _dynamic_reshape_indexer
    assert size_new == 1
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError:
  target: aten.view.default
  args[0]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float32,
      def inner_fn(index):
          i0, i1, i2, i3 = index
          tmp0 = ops.load(buf37, i3 + 49 * i2 + 2401 * i1 + 9604 * i0)
          tmp1 = ops.load(arg35_1, i3 + 49 * i2)
          tmp2 = ops.load(arg1_1, i1 + 4 * (tmp1))
          tmp3 = tmp0 + tmp2
          return tmp3
      ,
      ranges=[64, 4, 49, 49],
      origins={add_12}
    )
  ))
  args[1]: [64//s3, s3, 4, 49, 49]
```

the target shaps of ```view``` is ```[64//s3, s3, 4, 49, 49]```, and ```Sym(s3)``` is 64, see 

```
sym_size_16: Sym(s3) = torch.ops.aten.sym_size(arg34_1, 0)
floordiv_3: Sym(64//s3) = sym_size_13 // sym_size_16
view_33: f32[64//s3, 64//(64//s3), 4, 49, 49] = torch.ops.aten.view.default(add_12, [floordiv_3, sym_size_16, 4, sym_size_14, sym_size_14]);  add_12 = floordiv_3 = sym_size_16 = None
```
 
For the tail index of the new size is ```Sym(64//s3)```, it is not a number, we shouldn't directly compare it with ```1```.

Currently, I didn't find a simple test case to reproduce it, I just test it for the real model.

```
python -m torch.backends.xeon.run_cpu --core_list 0 --ncores_per_instance 1 benchmarks/dynamo/timm_models.py --performance --float32 -dcpu -n50 --inductor --only swin_base_patch4_window7_224 --batch_size 1 --threads 1
```

cc @ezyang @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: fix _dynamic_reshape_indexer issue when tail index is sym

[ghstack-poisoned]

ezyang commented:
My question is: is s3 actually legitimately dynamic, or is it actually static in the example? How come we're using size_hint as opposed to specializing?

ezyang commented:
Please try patching in https://github.com/pytorch/pytorch/pull/97428 and try again

ezyang commented:
ok, we'll have to bug @Chillee  some more about this when he's back from PTO.

In the meantime, I'm happy to land this if someone can answer my questions

XiaobingSuper commented:
> ok, we'll have to bug @Chillee some more about this when he's back from PTO.
> 
> In the meantime, I'm happy to land this if someone can answer my questions


For the model definition, the **s3** is a static shape which is from the mask: https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/swin_transformer.py#L259. So    should we use size_hint?

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1398,12 +1398,12 @@ def _dynamic_reshape_indexer(old_size, new_size):
         while stack_new:
             var, size_new = stack_new.pop()
-            assert size_new == 1
+            assert size_hint(size_new) == 1
```
Shouldn't these be guards?

XiaobingSuper committed:
Update on "inductor: fix _dynamic_reshape_indexer issue when tail index is sym"



For TIMM **swin_base_patch4_window7_224**  dynamic shape case, there has an error for ```view``` op:

```
  File "/home/xiaobing/pytorch-offical/torch/_inductor/lowering.py", line 229, in wrapped
    out = decomp_fn(*args, **kwargs)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/lowering.py", line 665, in view
    return TensorBox(View.create(x.data, sizes))
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1325, in create
    reindex = cls.dynamic_reshape_indexer(old_size, new_size)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1351, in dynamic_reshape_indexer
    reindex2 = cls._dynamic_reshape_indexer(flat, new_size)
  File "/home/xiaobing/pytorch-offical/torch/_inductor/ir.py", line 1406, in _dynamic_reshape_indexer
    assert size_new == 1
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError:
  target: aten.view.default
  args[0]: TensorBox(StorageBox(
    Pointwise(
      'cpu',
      torch.float32,
      def inner_fn(index):
          i0, i1, i2, i3 = index
          tmp0 = ops.load(buf37, i3 + 49 * i2 + 2401 * i1 + 9604 * i0)
          tmp1 = ops.load(arg35_1, i3 + 49 * i2)
          tmp2 = ops.load(arg1_1, i1 + 4 * (tmp1))
          tmp3 = tmp0 + tmp2
          return tmp3
      ,
      ranges=[64, 4, 49, 49],
      origins={add_12}
    )
  ))
  args[1]: [64//s3, s3, 4, 49, 49]
```

the target shaps of ```view``` is ```[64//s3, s3, 4, 49, 49]```, and ```Sym(s3)``` is 64, see 

```
sym_size_16: Sym(s3) = torch.ops.aten.sym_size(arg34_1, 0)
floordiv_3: Sym(64//s3) = sym_size_13 // sym_size_16
view_33: f32[64//s3, 64//(64//s3), 4, 49, 49] = torch.ops.aten.view.default(add_12, [floordiv_3, sym_size_16, 4, sym_size_14, sym_size_14]);  add_12 = floordiv_3 = sym_size_16 = None
```
 
For the tail index of the new size is ```Sym(64//s3)```, it is not a number, we shouldn't directly compare it with ```1```.

Currently, I didn't find a simple test case to reproduce it, I just test it for the real model.

```
python -m torch.backends.xeon.run_cpu --core_list 0 --ncores_per_instance 1 benchmarks/dynamo/timm_models.py --performance --float32 -dcpu -n50 --inductor --only swin_base_patch4_window7_224 --batch_size 1 --threads 1
```

cc ezyang soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang reviewed (approved):
okey dokey
